packages/framework/src/Component/LocalCache/LocalCacheFacade.php

```diff
@@ -0,0 +1,193 @@
+     */
+    protected function getNamespaceCache(string $namespace): ArrayAdapter
+    {
+        return $this->namespacesCache->getItem($namespace)->get();
+    }
+
+    /**
+     * @param string $namespace
+     * @param string $key
+     * @return bool
+     */
+    public function hasItem(string $namespace, string $key): bool
+    {
+        $namespace = $this->replaceNotAllowedCharactersInKey($namespace);
+
+        if ($this->hasNamespaceCache($namespace)) {
+            $key = $this->replaceNotAllowedCharactersInKey($key);
+
+            return $this->getNamespaceCache($namespace)->hasItem($key);
+        }
+
+        return false;
+    }
+
+    /**
+     * @param string $namespace
+     * @return bool
+     */
+    protected function hasNamespaceCache(string $namespace): bool
+    {
+        return $this->namespacesCache->hasItem($namespace);
+    }
+
+    public function reset(): void
+    {
+        $this->namespacesCache->reset();
+    }
+
+    /**
+     * @param string $namespace
+     * @param string $key
+     */
+    public function deleteItem(string $namespace, string $key): void
+    {
+        $namespace = $this->replaceNotAllowedCharactersInKey($namespace);
+
+        if (!$this->hasNamespaceCache($namespace)) {
+            return;
+        }
+
+        $key = $this->replaceNotAllowedCharactersInKey($key);
+        $namespaceCache = $this->getNamespaceCache($namespace);
+        $namespaceCache->deleteItem($key);
+    }
+
+    /**
+     * @param string $namespace
+     * @param string $key
+     * @param mixed $value
+     */
+    public function save(string $namespace, string $key, mixed $value): void
+    {
+        $namespace = $this->replaceNotAllowedCharactersInKey($namespace);
+        $key = $this->replaceNotAllowedCharactersInKey($key);
+        $namespaceCacheItem = $this->namespacesCache->getItem($namespace);
+
+        if (!$this->hasNamespaceCache($namespace)) {
+            $namespaceCacheItem->set(new ArrayAdapter(0, false));
+            $this->namespacesCache->save($namespaceCacheItem);
+        }
+
+        /** @var \Symfony\Component\Cache\Adapter\ArrayAdapter $namespaceCache */
+        $namespaceCache = $namespaceCacheItem->get();
+        $valueItem = $namespaceCache->getItem($key);
+        $valueItem->set($value);
+        $namespaceCache->save($valueItem);
+    }
+
+    /**
+     * @param string $key
+     * @return string
+     */
+    protected function replaceNotAllowedCharactersInKey(string $key): string
+    {
+        foreach (str_split(static::NOT_ALLOWED_CHARS) as $char) {
+            $key = str_replace($char, '~', $key);
+        }
+
+        return $key;
+    }
+
+    /**
+     * @param string $namespace
+     * @param callable $getValueCallback
+     * @param string|int|float|bool ...$keyParts List of variables for building cache key
+     * @return mixed
+     */
+    public function getOrSaveValue(string $namespace, callable $getValueCallback, ...$keyParts): mixed
```

Comment:
Some method have string key, some have keyParts, why are those two different?

Reply:
allows several different uses for different situations, it is up to the programmer which of the possible approaches he wants to use. I have left examples of different ways of use in concrete implementations.